### ISSUE TITLE
Unexpected reportIncompatibleMethodOverride when overriding a method with a Self-like type variable

### ISSUE THREAD

**Issue by Finistere** [microsoft/pyright#3280]
**Describe the bug**
Overriding a method having an argument with a `Self`-like type variable doesn't work when using the subclass as the concrete type.

**To Reproduce**
Code to typecheck:
```python
from __future__ import annotations

from typing import TypeVar

SelfA = TypeVar('SelfA', bound='A')


class A:
    def method(self: SelfA, other: SelfA) -> None:
        ...


# This doesn't work
class B(A):
    def method(self, other: B) -> None:
        ...


# This works
class C(A):
    def method(self, other: A) -> None:
        ...

```

**Expected behavior**
The example shouldn't raise any type errors.

**Screenshots or Code**
```
  test.py:15:9 - error: Method "method" overrides class "A" in an incompatible manner
    Parameter 2 type mismatch: base parameter is type "SelfA@method", override parameter is type "B"
      "A*" is incompatible with "B" (reportIncompatibleMethodOverride)
```

**VS Code extension or command-line**
I'm using `pyright`, version 1.1.234, from the command line with Python 3.9.6.

**Comment by erictraut** [comment]
Yes, this is intended behavior. The `reportIncompatibleMethodOverride` is enabled only in strict mode, and it's rather opinionated (i.e. strict) about what constitutes a "compatible override". In particular, the type of the method in the subclass must be assignable to the type of the method in the base class.

**Comment by Finistere** [comment]
Thank you for your response! Indeed, I didn't realize it was violating the Liskov substitution principle. I find it a bit confusing that PEP673 talks about `Self` in parameter types though: https://peps.python.org/pep-0673/#use-in-parameter-types. It seems to only make sense inside a `Protocol` to me, but the PEP example doesn't use one.
